### PR TITLE
Use implicit-hie cradle

### PR DESCRIPTION
**Pull request by Avi-D-coder**
Companion to #110

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
There's a issue with stack and `exe/Main.hs`

**Comment by Avi-D-coder** [comment]
Followup to this would be preferred cabal or stack config. I thought about doing this via `hie.yaml`, but that's the wrong approach it should be a LSP config, since `hie.yaml`s are often committed.

**Comment by fendor** [comment]
cabal-helper can implement this behaviour easily. Just filter for the project context, if there is none report an error, otherwise use the context

**Comment by Avi-D-coder** [comment]
In light of yet another [reddit post](https://www.reddit.com/r/haskell/comments/hcu5sf/mystified_by_hieyaml/?utm_source=share&utm_medium=web2x) by someone confused by `hie.yaml`.

### It's better then nothing
I would like to push for this to be merged, it does not always generate the exact right config, but it works for ever config a beginner would use. When it fails, the `hie-bios` no cradle matched error message is decipherable. When using conditionals or common stanzas, more advanced users will still have to write an `hie.yaml`,  but they will have a good error message and a tool `implicit-hie` to help them.

There are better solutions in the pipeline, but that should not stop us from improving user experience now. Let's not _let perfect be the enemy of good_.

**Comment by jneira** [comment]
Agree, i thought a more complete version would be ready sooner.
However i think that it would need a specific section in the readme, specifying in which known cases the implicit config will not work for sure.

File: cabal.project [hunk 1]
```diff
@@ -13,6 +13,16 @@ source-repository-package
     location: https://github.com/peti/cabal-plan
     tag: 894b76c0b6bf8f7d2f881431df1f13959a8fce87
 
+source-repository-package
+    type: git
+    location: https://github.com/mpickering/hie-bios
+    tag: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
```
Comment (fendor) [inline cabal.project:19]: Unneeded, afaict?

File: stack-8.10.1.yaml [hunk 2]
```diff
@@ -21,6 +21,11 @@ extra-deps:
 - optics-core-0.3
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack-8.10.1.yaml:28]: Two of these are unneeded, right?

File: stack-8.6.4.yaml [hunk 3]
```diff
@@ -48,6 +47,11 @@ extra-deps:
 # - shake-0.18.5
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack-8.6.4.yaml:54]: same as above

File: stack-8.6.5.yaml [hunk 4]
```diff
@@ -40,6 +39,11 @@ extra-deps:
 - semialign-1.1
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack-8.6.5.yaml:46]: same here

File: stack-8.8.2.yaml [hunk 5]
```diff
@@ -31,6 +30,11 @@ extra-deps:
 - semigroups-0.18.5
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack-8.8.2.yaml:37]: and here

File: stack-8.8.3.yaml [hunk 6]
```diff
@@ -30,6 +29,11 @@ extra-deps:
 - semigroups-0.18.5
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack-8.8.3.yaml:36]: and here

File: stack.yaml [hunk 7]
```diff
@@ -40,6 +39,11 @@ extra-deps:
 - semialign-1.1
 - github: wz1000/shake
   commit: fb3859dca2e54d1bbb2c873e68ed225fa179fbef
+- github: Avi-D-coder/implicit-hie-cradle
+  commit: 7de3cca59975db0e8a4280785b38d542a4a35c7a
+- github: mpickering/hie-bios
+  commit: a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
+- implicit-hie-0.1.0.0
```
Comment (fendor) [inline stack.yaml:46]: and here

**Comment by Avi-D-coder** [comment]
I'll rebase, when I get a chance hopefully tomorrow, very busy.

File: cabal.project (hunk 1 above)
Comment (Avi-D-coder) [reply cabal.project:19]: Needed for https://github.com/mpickering/hie-bios/commit/a852cffb0deb306cd1e1fe02aa50d50c77aeb15d

**Comment by Avi-D-coder** [comment]
@fendor I'm in the process of rebasing, but hie-bios has changed, how soon before the next hie-bios release. Unless we want hls to depend on an out of date, but still not published hie-bio commit we need to wait until hls is using a newer hie-bios.

**Comment by fendor** [comment]
@Avi-D-coder I want to hopefully do a release this week, but there are three pending PRs that need to be reviewed and merged. So, unfortunately, no promises.
Maybe we can do a minor release before the major release.

**Comment by alanz** [comment]
@fendor @Avi-D-coder what is holding this up?  I think it is important we get it in.

**Comment by fendor** [comment]
@alanz A hie-bios release that includes: https://github.com/mpickering/hie-bios/commit/a852cffb0deb306cd1e1fe02aa50d50c77aeb15d
I will get to it tomorrow.

**Comment by Avi-D-coder** [comment]
New version of hie-bios needs to be released, ghcide & hls need to use it, and then I need to find time to rebase and fix the wrapper.

**Comment by alanz** [comment]
Ok, so it is waiting, rather than abandoned. Thanks.

**Comment by Avi-D-coder** [comment]
Rebased at https://github.com/haskell/haskell-language-server/pull/186

**Pull request closed without merging**